src/Metalama.Patterns.NotifyPropertyChanged/Implementation/Natural/BuildAspectContext.cs

```diff
@@ -0,0 +1,451 @@
+﻿// Copyright (c) SharpCrafters s.r.o. See the LICENSE.md file in the root directory of this repository root for details.
+
+using Metalama.Framework.Aspects;
+using Metalama.Framework.Code;
+using Metalama.Framework.Code.Collections;
+using Metalama.Framework.Diagnostics;
+using Metalama.Framework.Engine.Diagnostics;
+using Metalama.Patterns.NotifyPropertyChanged.Metadata;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Metalama.Patterns.NotifyPropertyChanged.Implementation.Natural;
+
+[CompileTime]
+internal sealed partial class BuildAspectContext
+{
+    private static readonly string[] _onPropertyChangedMethodNames = { "OnPropertyChanged", "NotifyOfPropertyChange", "RaisePropertyChanged" };
+
+    private readonly Dictionary<IType, InpcInstrumentationKind> _inpcInstrumentationKindLookup = new();
+    private readonly Dictionary<IFieldOrProperty, bool> _validateFieldOrPropertyLookup = new();
+    private readonly Lazy<IMethod?> _baseOnPropertyChangedMethod;
+    private readonly Lazy<IMethod?> _baseOnChildPropertyChangedMethod;
+    private readonly Lazy<IMethod?> _baseOnUnmonitoredObservablePropertyChangedMethod;
```

Comment:
For clarity, it may be better to have a concept of `OnObservablePropertyChanged` instead of `OnUnmonitoredObservablePropertyChanged`, i.e. we remove the idea of the property being unmonitored. It makes the concept easier to understand and may have little performance impact.

Reply:
As I understand it, this seems to be mutually incompatible with the [comment below](https://github.com/postsharp/Metalama.Patterns/pull/23#discussion_r1337250378). I also question the assertion that there would be little performance impact, Suggest we discuss.

Reply:
Let open for now.